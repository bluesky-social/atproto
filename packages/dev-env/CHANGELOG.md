# @atproto/dev-env

## 0.2.17

### Patch Changes

- Updated dependencies [[`45352f9b`](https://github.com/bluesky-social/atproto/commit/45352f9b6d02aa405be94e9102424d983912ca5d)]:
  - @atproto/api@0.7.0
  - @atproto/bsky@0.0.17
  - @atproto/pds@0.3.5

## 0.2.16

### Patch Changes

- Updated dependencies [[`3c0ef382`](https://github.com/bluesky-social/atproto/commit/3c0ef382c12a413cc971ae47ffb341236c545f60), [`378fc613`](https://github.com/bluesky-social/atproto/commit/378fc6132f621ca517897c9467ed5bba134b3776)]:
  - @atproto/syntax@0.1.5
  - @atproto/api@0.6.24
  - @atproto/bsky@0.0.16
  - @atproto/lexicon@0.3.1
  - @atproto/pds@0.3.4
  - @atproto/xrpc-server@0.4.2

## 0.2.15

### Patch Changes

- Updated dependencies [[`e1b5f253`](https://github.com/bluesky-social/atproto/commit/e1b5f2537a5ba4d8b951a741269b604856028ae5)]:
  - @atproto/crypto@0.3.0
  - @atproto/xrpc-server@0.4.1
  - @atproto/bsky@0.0.15
  - @atproto/identity@0.3.2
  - @atproto/pds@0.3.3
  - @atproto/api@0.6.23

## 0.2.14

### Patch Changes

- Updated dependencies [[`772736a0`](https://github.com/bluesky-social/atproto/commit/772736a01081f39504e1b19a1b3687783bb78f07)]:
  - @atproto/api@0.6.23
  - @atproto/bsky@0.0.14
  - @atproto/pds@0.3.2

## 0.2.13

### Patch Changes

- [#1788](https://github.com/bluesky-social/atproto/pull/1788) [`84e2d4d2`](https://github.com/bluesky-social/atproto/commit/84e2d4d2b6694f344d80c18672c78b650189d423) Thanks [@bnewbold](https://github.com/bnewbold)! - update license to "MIT or Apache2"

- Updated dependencies [[`ce49743d`](https://github.com/bluesky-social/atproto/commit/ce49743d7f8800d33116b88001d7b512553c2c89), [`84e2d4d2`](https://github.com/bluesky-social/atproto/commit/84e2d4d2b6694f344d80c18672c78b650189d423)]:
  - @atproto/lexicon@0.3.0
  - @atproto/xrpc-server@0.4.0
  - @atproto/common-web@0.2.3
  - @atproto/identity@0.3.1
  - @atproto/crypto@0.2.3
  - @atproto/syntax@0.1.4
  - @atproto/bsky@0.0.13
  - @atproto/api@0.6.22
  - @atproto/pds@0.3.1

## 0.2.12

### Patch Changes

- Updated dependencies [[`9c98a5ba`](https://github.com/bluesky-social/atproto/commit/9c98a5baaf503b02238a6afe4f6e2b79c5181693), [`bb039d8e`](https://github.com/bluesky-social/atproto/commit/bb039d8e4ce5b7f70c4f3e86d1327e210ef24dc3), [`35d108ce`](https://github.com/bluesky-social/atproto/commit/35d108ce94866ce1b3d147cd0620a0ba1c4ebcd7)]:
  - @atproto/api@0.6.21
  - @atproto/identity@0.3.0
  - @atproto/common-web@0.2.2
  - @atproto/bsky@0.0.12
  - @atproto/pds@0.3.0
  - @atproto/lexicon@0.2.3
  - @atproto/syntax@0.1.3
  - @atproto/xrpc-server@0.3.3

## 0.2.11

### Patch Changes

- [#1568](https://github.com/bluesky-social/atproto/pull/1568) [`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b) Thanks [@dholms](https://github.com/dholms)! - Added email verification and update flows

- Updated dependencies [[`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b), [`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b)]:
  - @atproto/api@0.6.20
  - @atproto/pds@0.1.20
  - @atproto/common-web@0.2.1
  - @atproto/bsky@0.0.11
  - @atproto/identity@0.2.1
  - @atproto/syntax@0.1.2
  - @atproto/xrpc-server@0.3.2

## 0.2.10

### Patch Changes

- Updated dependencies [[`35b616cd`](https://github.com/bluesky-social/atproto/commit/35b616cd82232879937afc88d3f77d20c6395276)]:
  - @atproto/api@0.6.19
  - @atproto/bsky@0.0.10
  - @atproto/pds@0.1.19

## 0.2.9

### Patch Changes

- Updated dependencies [[`2ce8a11b`](https://github.com/bluesky-social/atproto/commit/2ce8a11b8daf5d39027488c5dde8c47b0eb937bf)]:
  - @atproto/api@0.6.18
  - @atproto/bsky@0.0.9
  - @atproto/pds@0.1.18

## 0.2.8

### Patch Changes

- Updated dependencies [[`d96f7d9b`](https://github.com/bluesky-social/atproto/commit/d96f7d9b84c6fbab9711059c8584a77d892dcedd)]:
  - @atproto/bsky@0.0.8
  - @atproto/api@0.6.17
  - @atproto/pds@0.1.17

## 0.2.7

### Patch Changes

- Updated dependencies [[`56e2cf89`](https://github.com/bluesky-social/atproto/commit/56e2cf8999f6d7522529a9be8652c47545f82242)]:
  - @atproto/api@0.6.16
  - @atproto/bsky@0.0.7
  - @atproto/pds@0.1.16

## 0.2.6

### Patch Changes

- Updated dependencies [[`2cc329f2`](https://github.com/bluesky-social/atproto/commit/2cc329f26547217dd94b6bb11ee590d707cbd14f)]:
  - @atproto/api@0.6.15
  - @atproto/bsky@0.0.6
  - @atproto/pds@0.1.15

## 0.2.5

### Patch Changes

- Updated dependencies [[`b1dc3555`](https://github.com/bluesky-social/atproto/commit/b1dc355504f9f2e047093dc56682b8034518cf80)]:
  - @atproto/syntax@0.1.1
  - @atproto/api@0.6.14
  - @atproto/bsky@0.0.5
  - @atproto/pds@0.1.14
  - @atproto/xrpc-server@0.3.1

## 0.2.4

### Patch Changes

- Updated dependencies [[`3877210e`](https://github.com/bluesky-social/atproto/commit/3877210e7fb3c76dfb1a11eb9ba3f18426301d9f)]:
  - @atproto/api@0.6.13
  - @atproto/bsky@0.0.4
  - @atproto/pds@0.1.13
