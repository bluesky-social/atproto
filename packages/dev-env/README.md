# @atproto/dev-env: Local Developer Environment

A command-line application for developers to construct and manage development environments.

[![NPM](https://img.shields.io/npm/v/@atproto/dev-env)](https://www.npmjs.com/package/@atproto/dev-env)
[![Github CI Status](https://github.com/bluesky-social/atproto/actions/workflows/repo.yaml/badge.svg)](https://github.com/bluesky-social/atproto/actions/workflows/repo.yaml)

## REPL API

The following methods are available in the REPL.

### `status()`

List the currently active servers.

### `startPds(port?: number)`

Create a new PDS instance. Data is stored in memory.

### `stop(port: number)`

Stop the server at the given port.

### `mkuser(handle: string, pdsPort?: number)`

Create a new user.

### `user(handle: string): ServiceClient`

Get the `ServiceClient` for the given user.

## License

This project is dual-licensed under MIT and Apache 2.0 terms:

- MIT license ([LICENSE-MIT.txt](https://github.com/bluesky-social/atproto/blob/main/LICENSE-MIT.txt) or http://opensource.org/licenses/MIT)
- Apache License, Version 2.0, ([LICENSE-APACHE.txt](https://github.com/bluesky-social/atproto/blob/main/LICENSE-APACHE.txt) or http://www.apache.org/licenses/LICENSE-2.0)

Downstream projects and end users may chose either license individually, or both together, at their discretion. The motivation for this dual-licensing is the additional software patent assurance provided by Apache 2.0.
