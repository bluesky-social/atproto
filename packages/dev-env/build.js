const { copy } = require('esbuild-plugin-copy')

require('esbuild')
  .build({
    logLevel: 'info',
    entryPoints: ['src/index.ts', 'src/cli.ts'],
    bundle: true,
    sourcemap: true,
    outdir: 'dist',
    platform: 'node',
    external: [
      'better-sqlite3',
      // Referenced in pg driver, but optional and we don't use it
      'pg-native',
      'sharp',
    ],
    plugins: [
      copy({
        assets: {
          from: ['../pds/src/mailer/templates/**/*'],
          to: ['./templates'],
          keepStructure: true,
        },
      }),
    ],
  })
  .catch(() => process.exit(1))
