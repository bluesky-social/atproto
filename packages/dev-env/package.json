{
  "name": "@atproto/dev-env",
  "version": "0.1.0",
  "main": "src/test-env.ts",
  "bin": "dist/cli.js",
  "license": "MIT",
  "repository": "git@github.com:bluesky-social/atproto.git",
  "scripts": {
    "build": "node ./build.js",
    "postbuild": "tsc --build tsconfig.build.json",
    "start": "node dist/cli.js",
    "prettier": "prettier --check src/",
    "prettier:fix": "prettier --write src/",
    "lint": "eslint . --ext .ts,.tsx",
    "lint:fix": "yarn lint --fix",
    "verify": "run-p prettier lint",
    "verify:fix": "yarn prettier:fix && yarn lint:fix"
  },
  "dependencies": {
    "@atproto/api": "*",
    "@atproto/bsky": "*",
    "@atproto/crypto": "*",
    "@atproto/did-resolver": "*",
    "@atproto/pds": "*",
    "@atproto/uri": "*",
    "@did-plc/lib": "^0.0.1",
    "@did-plc/server": "^0.0.1",
    "better-sqlite3": "^7.6.2",
    "chalk": "^5.0.1",
    "dotenv": "^16.0.1",
    "get-port": "^6.1.2",
    "sharp": "^0.31.2"
  },
  "devDependencies": {
    "ts-node": "^10.8.1"
  }
}
