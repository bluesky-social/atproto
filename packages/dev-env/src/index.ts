import http from 'http'
import chalk from 'chalk'
import crytpo from 'crypto'
import PDSServer, {
  Database as PDSDatabase,
  MemoryBlobStore,
  ServerConfig as PDSServerConfig,
  AppContext as PDSContext,
} from '@atproto/pds'
import * as bsky from '@atproto/bsky'
import * as plc from '@did-plc/lib'
import { PlcServer, Database as PlcDatabase } from '@did-plc/server'
import * as crypto from '@atproto/crypto'
import AtpAgent from '@atproto/api'
import { ServerType, ServerConfig, StartParams, PORTS } from './types.js'
import { HOUR } from '@atproto/common'
import { mockNetworkUtilities } from './test-env'

export * from './test-env'

interface Startable {
  start(): Promise<http.Server>
}

interface Destroyable {
  destroy(): Promise<void>
}

export class DevEnvServer {
  inst?: Destroyable

  constructor(
    private env: DevEnv,
    public type: ServerType,
    public port: number,
  ) {}

  get name() {
    return {
      [ServerType.PersonalDataServer]: 'ðŸŒž Personal Data server',
      [ServerType.DidPlaceholder]: 'ðŸ‘¤ DID Placeholder server',
      [ServerType.BskyAppView]: 'ðŸŒ€ Bsky App View server',
    }[this.type]
  }

  get description() {
    return `[${chalk.bold(this.port)}] ${this.name}`
  }

  get url() {
    return `http://localhost:${this.port}`
  }

  get ctx(): PDSContext | undefined {
    if (this.inst instanceof PDSServer) {
      return this.inst.ctx
    }
  }

  async start() {
    if (this.inst) {
      throw new Error('Already started')
    }

    const startServer = async (server: Startable): Promise<void> => {
      try {
        await server.start()
        console.log(`${this.description} started ${chalk.gray(this.url)}`)
      } catch (err) {
        console.log(`${this.description} failed to start:`, err)
      }
    }

    switch (this.type) {
      case ServerType.PersonalDataServer: {
        if (!this.env.plcUrl) {
          throw new Error('Must be running a PLC server to start a PDS')
        }

        const db = await PDSDatabase.memory()
        await db.migrateToLatestOrThrow()
        const keypair = await crypto.EcdsaKeypair.create()

        const blobstore = new MemoryBlobStore()

        const plcClient = new plc.Client(this.env.plcUrl)
        const serverDid = await plcClient.createDid({
          signingKey: keypair.did(),
          rotationKeys: [keypair.did()],
          handle: 'localhost',
          pds: `http://localhost:${this.port}`,
          signer: keypair,
        })

        const pds = PDSServer.create({
          db,
          blobstore,
          repoSigningKey: keypair,
          plcRotationKey: keypair,
          config: new PDSServerConfig({
            debugMode: true,
            version: '0.0.0',
            scheme: 'http',
            hostname: 'localhost',
            port: this.port,
            didPlcUrl: this.env.plcUrl,
            serverDid,
            recoveryKey: keypair.did(),
            jwtSecret: crytpo.randomBytes(8).toString('base64'),
            availableUserDomains: ['.test'],
            appUrlPasswordReset: 'app://password-reset',
            // @TODO setup ethereal.email creds and set emailSmtpUrl here
            emailNoReplyAddress: 'noreply@blueskyweb.xyz',
            adminPassword: 'password',
            inviteRequired: false,
            userInviteInterval: null,
            imgUriSalt: '9dd04221f5755bce5f55f47464c27e1e',
            imgUriKey:
              'f23ecd142835025f42c3db2cf25dd813956c178392760256211f9d315f8ab4d8',
            privacyPolicyUrl: 'https://example.com/privacy',
            termsOfServiceUrl: 'https://example.com/tos',
            labelerDid: 'did:example:labeler',
            labelerKeywords: {},
            maxSubscriptionBuffer: 200,
            repoBackfillLimitMs: HOUR,
            appViewRepoProvider: process.env.APP_VIEW_REPO_PROVIDER,
          }),
        })
        await startServer(pds)
        this.inst = pds
        break
      }
      case ServerType.DidPlaceholder: {
        const db = PlcDatabase.mock()
        const plcServer = PlcServer.create({ db, port: this.port })
        await startServer(plcServer)
        this.inst = plcServer
        break
      }
      case ServerType.BskyAppView: {
        if (!this.env.pdsUrl || !this.env.plcUrl) {
          throw new Error(
            'Must be running a PDS and PLC servers to start AppView',
          )
        }

        const config = new bsky.ServerConfig({
          version: '0.0.0',
          didPlcUrl: this.env.plcUrl,
          publicUrl: 'https://bsky.public.url',
          imgUriSalt: '9dd04221f5755bce5f55f47464c27e1e',
          imgUriKey:
            'f23ecd142835025f42c3db2cf25dd813956c178392760256211f9d315f8ab4d8',
          adminPassword: 'password',
          labelerDid: 'did:example:labeler',
          dbPostgresUrl: process.env.DB_POSTGRES_URL || '',
          dbPostgresSchema: process.env.DB_POSTGRES_SCHEMA,
          repoProvider: this.env.pdsUrl.replace('http://', 'ws://'),
          port: this.port,
          labelerKeywords: {},
        })

        const db = bsky.Database.postgres({
          url: config.dbPostgresUrl,
          schema: config.dbPostgresSchema,
        })

        await db.migrateToLatestOrThrow()

        const server = bsky.BskyAppView.create({ db, config })
        await startServer(server)
        this.inst = server
        break
      }
      default:
        throw new Error(`Unsupported server type: ${this.type}`)
    }
  }

  async close() {
    if (this.inst) {
      console.log(`Closing ${this.description}`)
      await this.inst.destroy()
    }
  }

  getClient(): AtpAgent {
    return new AtpAgent({ service: `http://localhost:${this.port}` })
  }
}

export class DevEnv {
  plcUrl: string | undefined
  pdsUrl: string | undefined
  servers: Map<number, DevEnvServer> = new Map()

  static async create(params: StartParams): Promise<DevEnv> {
    const devEnv = new DevEnv()
    for (const cfg of params.servers || []) {
      await devEnv.add(cfg)
    }
    return devEnv
  }

  async add(cfg: ServerConfig) {
    if (this.servers.has(cfg.port)) {
      throw new Error(`Port ${cfg.port} is in use`)
    } else if (cfg.type === ServerType.DidPlaceholder && this.plcUrl) {
      throw new Error('There should only be one plc server')
    }
    const server = new DevEnvServer(this, cfg.type, cfg.port)
    await server.start()
    this.servers.set(cfg.port, server)
    if (cfg.type === ServerType.DidPlaceholder) {
      this.plcUrl = `http://localhost:${cfg.port}`
    }
    if (cfg.type === ServerType.PersonalDataServer) {
      this.pdsUrl = `http://localhost:${cfg.port}`
    }
    if (cfg.type === ServerType.BskyAppView) {
      const pds = this.servers.get(PORTS[ServerType.PersonalDataServer])
      if (pds) {
        await mockNetworkUtilities({
          port: pds.port,
          url: pds.url,
          close: pds.close,
          ctx: pds.ctx!,
        })
      }
    }
  }

  async remove(server: number | DevEnvServer) {
    const port = typeof server === 'number' ? server : server.port
    const inst = this.servers.get(port)
    if (inst) {
      await inst.close()
      this.servers.delete(port)
    }
  }

  async shutdown() {
    for (const server of this.servers.values()) {
      await server.close()
    }
  }

  hasType(type: ServerType) {
    for (const s of this.servers.values()) {
      if (s.type === type) {
        return true
      }
    }
    return false
  }

  listOfType(type: ServerType): DevEnvServer[] {
    return Array.from(this.servers.values()).filter((s) => s.type === type)
  }
}
