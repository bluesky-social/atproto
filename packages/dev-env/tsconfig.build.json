{
  "extends": "../../tsconfig/node.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "noPropertyAccessFromIndexSignature": false
  },
  "include": ["./src"]
}
