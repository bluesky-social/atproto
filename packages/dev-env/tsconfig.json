{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist", // Your outDir,
    "emitDeclarationOnly": true
  },
  "module": "esnext",
  "include": ["./src","__tests__/**/**.ts"],
  "references": [
    { "path": "../api/tsconfig.build.json" },
    { "path": "../bsky/tsconfig.build.json" },
    { "path": "../crypto/tsconfig.build.json" },
    { "path": "../did-resolver/tsconfig.build.json" },
    { "path": "../pds/tsconfig.json" },
    { "path": "../uri/tsconfig.build.json" },
  ]
}