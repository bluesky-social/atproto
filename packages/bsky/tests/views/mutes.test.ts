import AtpAgent from '@atproto/api'
import { TestNetwork, SeedClient } from '@atproto/dev-env'
import { forSnapshot, paginateAll } from '../_util'
import basicSeed from '../seeds/basic'
import usersBulkSeed from '../seeds/users-bulk'

describe('mute views', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let sc: SeedClient
  let alice: string
  let bob: string
  let carol: string
  let dan: string

  let mutes: string[]

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_mutes',
    })
    agent = network.bsky.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    await usersBulkSeed(sc, 10)
    alice = sc.dids.alice
    bob = sc.dids.bob
    carol = sc.dids.carol
    dan = sc.dids.dan
    mutes = [
      bob,
      carol,
      'aliya-hodkiewicz.test',
      'adrienne49.test',
      'jeffrey-sawayn87.test',
      'nicolas-krajcik10.test',
      'magnus53.test',
      'elta48.test',
    ]
    await network.processAll()
    for (const did of mutes) {
      await agent.api.app.bsky.graph.muteActor(
        { actor: did },
        {
          headers: await network.serviceHeaders(alice),
          encoding: 'application/json',
        },
      )
    }
  })

  afterAll(async () => {
    await network.close()
  })

  it('flags mutes in threads', async () => {
    const res = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[alice][1].ref.uriStr },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(forSnapshot(res.data.thread)).toMatchSnapshot()
  })

  it('does not show reposted content from a muted account in author feed', async () => {
    await sc.repost(dan, sc.posts[bob][0].ref)
    await sc.repost(dan, sc.posts[bob][1].ref)
    await network.processAll()

    const res = await agent.api.app.bsky.feed.getAuthorFeed(
      { actor: dan },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(
      res.data.feed.some((post) => [bob, carol].includes(post.post.author.did)),
    ).toBe(false)
  })

  it('removes content from muted users on getTimeline', async () => {
    const res = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(
      res.data.feed.some((post) => [bob, carol].includes(post.post.author.did)),
    ).toBe(false)
  })

  it('removes content from muted users on getListFeed', async () => {
    const listRef = await sc.createList(bob, 'test list', 'curate')
    await sc.addToList(alice, bob, listRef)
    await sc.addToList(alice, carol, listRef)
    await sc.addToList(alice, dan, listRef)
    const res = await agent.api.app.bsky.feed.getListFeed(
      { list: listRef.uriStr },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(
      res.data.feed.some((post) => [bob, carol].includes(post.post.author.did)),
    ).toBe(false)
  })

  it('returns mute status on getProfile', async () => {
    const res = await agent.api.app.bsky.actor.getProfile(
      { actor: bob },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(res.data.viewer?.muted).toBe(true)
  })

  it('returns mute status on getProfiles', async () => {
    const res = await agent.api.app.bsky.actor.getProfiles(
      { actors: [bob, carol, dan] },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(res.data.profiles[0].viewer?.muted).toBe(true)
    expect(res.data.profiles[1].viewer?.muted).toBe(true)
    expect(res.data.profiles[2].viewer?.muted).toBe(false)
  })

  it('does not return notifs for muted accounts', async () => {
    const res = await agent.api.app.bsky.notification.listNotifications(
      {
        limit: 100,
      },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(
      res.data.notifications.some((notif) =>
        [bob, carol].includes(notif.author.did),
      ),
    ).toBeFalsy()
  })

  it('flags muted accounts in get suggestions', async () => {
    // unfollow so they _would_ show up in suggestions if not for mute
    await sc.unfollow(alice, bob)
    await sc.unfollow(alice, carol)

    await network.processAll()

    const res = await agent.api.app.bsky.actor.getSuggestions(
      {
        limit: 100,
      },
      { headers: await network.serviceHeaders(alice) },
    )
    for (const actor of res.data.actors) {
      if (mutes.includes(actor.did) || mutes.includes(actor.handle)) {
        expect(actor.viewer?.muted).toBe(true)
      } else {
        expect(actor.viewer?.muted).toBe(false)
      }
    }
  })

  it('fetches mutes for the logged-in user.', async () => {
    const { data: view } = await agent.api.app.bsky.graph.getMutes(
      {},
      { headers: await network.serviceHeaders(alice) },
    )
    expect(forSnapshot(view.mutes)).toMatchSnapshot()
  })

  it('paginates.', async () => {
    const results = (results) => results.flatMap((res) => res.mutes)
    const paginator = async (cursor?: string) => {
      const { data: view } = await agent.api.app.bsky.graph.getMutes(
        { cursor, limit: 2 },
        { headers: await network.serviceHeaders(alice) },
      )
      return view
    }

    const paginatedAll = await paginateAll(paginator)
    paginatedAll.forEach((res) =>
      expect(res.mutes.length).toBeLessThanOrEqual(2),
    )

    const full = await agent.api.app.bsky.graph.getMutes(
      {},
      { headers: await network.serviceHeaders(alice) },
    )

    expect(full.data.mutes.length).toEqual(8)
    expect(results(paginatedAll)).toEqual(results([full.data]))
  })

  it('removes mute.', async () => {
    const { data: initial } = await agent.api.app.bsky.graph.getMutes(
      {},
      { headers: await network.serviceHeaders(alice) },
    )
    expect(initial.mutes.length).toEqual(8)
    expect(initial.mutes.map((m) => m.handle)).toContain('elta48.test')

    await agent.api.app.bsky.graph.unmuteActor(
      { actor: sc.dids['elta48.test'] },
      {
        headers: await network.serviceHeaders(alice),
        encoding: 'application/json',
      },
    )

    const { data: final } = await agent.api.app.bsky.graph.getMutes(
      {},
      { headers: await network.serviceHeaders(alice) },
    )
    expect(final.mutes.length).toEqual(7)
    expect(final.mutes.map((m) => m.handle)).not.toContain('elta48.test')

    await agent.api.app.bsky.graph.muteActor(
      { actor: sc.dids['elta48.test'] },
      {
        headers: await network.serviceHeaders(alice),
        encoding: 'application/json',
      },
    )
  })

  it('does not allow muting self.', async () => {
    const promise = agent.api.app.bsky.graph.muteActor(
      { actor: alice },
      {
        headers: await network.serviceHeaders(alice),
        encoding: 'application/json',
      },
    )
    await expect(promise).rejects.toThrow('Cannot mute oneself')
  })
})
