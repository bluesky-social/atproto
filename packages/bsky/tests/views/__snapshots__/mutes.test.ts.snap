// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mute views fetches mutes for the logged-in user. 1`] = `
Array [
  Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
    "did": "user(0)",
    "displayName": "Dr. Lowell DuBuque",
    "handle": "elta48.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "muted": true,
    },
  },
  Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(3)/cids(0)@jpeg",
    "did": "user(2)",
    "displayName": "Sally Funk",
    "handle": "magnus53.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "muted": true,
    },
  },
  Object {
    "did": "user(4)",
    "handle": "nicolas-krajcik10.test",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "muted": true,
    },
  },
  Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(6)/cids(0)@jpeg",
    "did": "user(5)",
    "displayName": "Patrick Sawayn",
    "handle": "jeffrey-sawayn87.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "muted": true,
    },
  },
  Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(8)/cids(0)@jpeg",
    "did": "user(7)",
    "displayName": "Kim Streich",
    "handle": "adrienne49.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "muted": true,
    },
  },
  Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(10)/cids(0)@jpeg",
    "did": "user(9)",
    "displayName": "Carlton Abernathy IV",
    "handle": "aliya-hodkiewicz.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "muted": true,
    },
  },
  Object {
    "did": "user(11)",
    "handle": "carol.test",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "followedBy": "record(0)",
      "muted": true,
    },
  },
  Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(13)/cids(0)@jpeg",
    "description": "hi im bob label_me",
    "did": "user(12)",
    "displayName": "bobby",
    "handle": "bob.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "viewer": Object {
      "blockedBy": false,
      "followedBy": "record(1)",
      "muted": true,
    },
  },
]
`;

exports[`mute views flags mutes in threads 1`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 3,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000000Z",
      "text": "again",
    },
    "replyCount": 2,
    "repostCount": 1,
    "uri": "record(0)",
    "viewer": Object {},
  },
  "replies": Array [
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "did": "user(2)",
          "handle": "carol.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "followedBy": "record(4)",
            "following": "record(3)",
            "muted": true,
          },
        },
        "cid": "cids(3)",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "of course",
        },
        "replyCount": 0,
        "repostCount": 0,
        "uri": "record(2)",
        "viewer": Object {},
      },
    },
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "avatar": "https://bsky.public.url/img/avatar/plain/user(4)/cids(1)@jpeg",
          "did": "user(3)",
          "displayName": "bobby",
          "handle": "bob.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "followedBy": "record(7)",
            "following": "record(6)",
            "muted": true,
          },
        },
        "cid": "cids(4)",
        "embed": Object {
          "$type": "app.bsky.embed.images#view",
          "images": Array [
            Object {
              "alt": "tests/sample-img/key-landscape-small.jpg",
              "fullsize": "https://bsky.public.url/img/feed_fullsize/plain/user(4)/cids(5)@jpeg",
              "thumb": "https://bsky.public.url/img/feed_thumbnail/plain/user(4)/cids(5)@jpeg",
            },
          ],
        },
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(4)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "did:example:labeler",
            "uri": "record(5)",
            "val": "test-label",
          },
          Object {
            "cid": "cids(4)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "did:example:labeler",
            "uri": "record(5)",
            "val": "test-label-2",
          },
        ],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "embed": Object {
            "$type": "app.bsky.embed.images",
            "images": Array [
              Object {
                "alt": "tests/sample-img/key-landscape-small.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(5)",
                  },
                  "size": 4114,
                },
              },
            ],
          },
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "hear that label_me label_me_2",
        },
        "replyCount": 1,
        "repostCount": 0,
        "uri": "record(5)",
        "viewer": Object {},
      },
    },
  ],
}
`;
